src/main/java/com/salesforce/storm/spout/dynamic/MessageBus.java

```diff
@@ -0,0 +1,177 @@
+     */
+    public MessageBus(final MessageBuffer messageBuffer) {
+        this.messageBuffer = messageBuffer;
+    }
+
+    // SpoutMessageBus Interface methods.
+
+    @Override
+    public Optional<Throwable> getErrors() {
+        // Poll is non-blocking.
+        return Optional.ofNullable(reportedErrorsQueue.poll());
+    }
+
+    @Override
+    public Optional<Message> nextMessage() {
+        return Optional.ofNullable(messageBuffer.poll());
+    }
+
+    @Override
+    public void ack(final MessageId id) {
+        // TODO this actually isn't thread safe :(
+        if (!ackedTuplesQueue.containsKey(id.getSrcVirtualSpoutId())) {
+            logger.warn("Acking tuple for unknown consumer");
+            return;
+        }
+
+        ackedTuplesQueue.get(id.getSrcVirtualSpoutId()).add(id);
+    }
+
+    @Override
+    public void fail(final MessageId id) {
+        // TODO this actually isn't thread safe :(
+        if (!failedTuplesQueue.containsKey(id.getSrcVirtualSpoutId())) {
+            logger.warn("Failing tuple for unknown consumer");
+            return;
+        }
+
+        failedTuplesQueue.get(id.getSrcVirtualSpoutId()).add(id);
+    }
+
+    // VirtualSpoutMessageBus interface methods.
+
+    @Override
+    public void registerVirtualSpout(final VirtualSpoutIdentifier virtualSpoutIdentifier) {
+        messageBuffer.addVirtualSpoutId(virtualSpoutIdentifier);
+        ackedTuplesQueue.put(virtualSpoutIdentifier, new ConcurrentLinkedQueue<>());
+        failedTuplesQueue.put(virtualSpoutIdentifier, new ConcurrentLinkedQueue<>());
+    }
+
+    @Override
+    public void publishMessage(final Message message) throws InterruptedException {
+        messageBuffer.put(message);
+    }
+
+    @Override
+    public int messageSize() {
+        return messageBuffer.size();
+    }
+
+    @Override
+    public void publishError(final Throwable throwable) {
+        reportedErrorsQueue.add(throwable);
+    }
+
+    @Override
+    public Optional<MessageId> getAckedMessage(final VirtualSpoutIdentifier virtualSpoutIdentifier) {
+        final MessageId id = ackedTuplesQueue.get(virtualSpoutIdentifier).poll();
+        return Optional.ofNullable(id);
+    }
+
+    @Override
+    public Optional<MessageId> getFailedMessage(final VirtualSpoutIdentifier virtualSpoutIdentifier) {
+        final MessageId id = failedTuplesQueue.get(virtualSpoutIdentifier).poll();
+        return Optional.ofNullable(id);
+    }
+
+    @Override
+    public void unregisterVirtualSpout(final VirtualSpoutIdentifier virtualSpoutIdentifier) {
+        messageBuffer.removeVirtualSpoutId(virtualSpoutIdentifier);
+        ackedTuplesQueue.remove(virtualSpoutIdentifier);
+        failedTuplesQueue.remove(virtualSpoutIdentifier);
+    }
+
+    // Additional Helper methods not exposed via interfaces.
+
+    /**
+     * @return Count of acked messageIds that exist within the bus.
+     */
+    public int ackSize() {
```

Comment:
IMO these are only public because of a poorly written and located test. This should be package only. Same with `failSize()`

Reply:
I 100% agree with you, but I'm unsure how to refactor the tests in a way to achieve the same behavior.  I could change them to instead Thread.sleep() for a period but that seems more likely to flap.

Reply:
Going to merge this in, but if you have ideas for this comment up and I'll create a follow up PR and adjust it.